### PR TITLE
Create input refactor

### PR DESCRIPTION
**Pull request by freesig**
### TODO:
- [ ] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

### CODE REVIEW

File: crates/holochain_integrity_types/src/entry.rs [hunk 1]
```diff
@@ -64,6 +75,32 @@ pub enum Entry {
...
+    pub trait Sealed {}
+}
+
+impl AppEntryData for AppEntryBytes {}
+impl sealed::Sealed for AppEntryBytes {}
+impl AppEntryData for AppEntry {}
+impl sealed::Sealed for AppEntry {}
+
+#[derive(PartialEq, Clone, Debug, serde::Serialize, serde::Deserialize)]
+/// The data needed to build an [`Entry`] and it's [`Header`](crate::header::Header).
```
Comment (maackle) [inline crates/holochain_integrity_types/src/entry.rs:94]:
```suggestion
/// The data needed to build an [`Entry`] and its [`Action`](crate::action::Action).
```

File: crates/holochain_types/src/record.rs [hunk 2]
```diff
@@ -445,6 +446,51 @@ impl WireRecord {
     }
 }
 
+/// Helper trait for the [`RecordBuilder`].
+pub trait RecordBuilderHelper {
+    /// Build this element into it's parts for constructing an [`Element`].
```
Comment (maackle) [inline crates/holochain_types/src/record.rs:451]:
```suggestion
    /// Produce the parts needed to construct a [`Record`].
```

File: crates/holochain/src/test_utils.rs [hunk 3]
```diff
@@ -817,3 +818,42 @@ pub async fn force_publish_dht_ops(
     publish_trigger.trigger(&"force_publish_dht_ops");
     Ok(())
 }
+
+/// Helper trait for building [`CreateInput`] for tests.
+pub trait CreateInputBuilder {
```
Comment (maackle) [inline crates/holochain/src/test_utils.rs:823]:
```suggestion
pub trait CreateInputExt {
```

File: crates/holochain_integrity_types/src/entry.rs [hunk 4]
```diff
@@ -45,17 +47,26 @@ impl From<EntryHashed> for Entry {
 }
 
 /// Structure holding the entry portion of a chain record.
+pub type Entry = EntryImpl<AppEntryBytes>;
+
+/// Data for building a [`Record`](crate::record::Record).
+pub type RecordBuilder = EntryImpl<AppEntry>;
```
Comment (maackle) [inline crates/holochain_integrity_types/src/entry.rs:53]: What do you think about calling this something else? My objections is that this doesn't really evoke [the Builder pattern](https://rust-unofficial.github.io/patterns/patterns/creational/builder.html). `build()` doesn't even produce the thing that this is supposed to be a Builder for (a Record), it just produces some intermediate state, which has to be built up into a record elsewhere.

```suggestion
/// All entry-specific data needed to build a [`Record`](crate::record::Record).
pub type EntryPlus = EntryImpl<AppEntry>;
```

File: crates/holochain_types/src/record.rs [hunk 5]
```diff
@@ -445,6 +446,51 @@ impl WireRecord {
...
+    fn build(self) -> (EntryDefLocation, EntryVisibility, Entry);
+}
+
+impl RecordBuilderHelper for RecordBuilder {
+    fn build(self) -> (EntryDefLocation, EntryVisibility, Entry) {
```
Comment (maackle) [inline crates/holochain_types/src/record.rs:456]: Does this really need to be a trait? Could it just be a method directly on `RecordBuilder` (or whatever we call it)?

Also for the same reason I'd like this not to be called a Builder, I'd like this method to not be called `build`.

```suggestion
impl RecordBuilderHelper for RecordBuilder {
    fn into_parts(self) -> (EntryDefLocation, EntryVisibility, Entry) {
```

**Review by maackle — CHANGES_REQUESTED** [review]
Looks good, but now I have some concerns about things being called Builders.

**Review by jost-s — APPROVED** [review]
LGTM

**Review by maackle — APPROVED** [review]

**Pull request closed without merging**